### PR TITLE
Updates TypeScript and export declarations

### PR DESCRIPTION
**Pull request by giacomorebonato**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [*] I acknowledge that all my contributions will be made under the project's license.

### Description

I like to use TypeScript in my projects and I noticed that this library has been on the way to be fully ported to it in the last months.
I did this exploration to see if I was able to export at least the type declaration files.

~I wasn't able to make the `typedoc` script working as I am not confident with the fork made and with the library in general; that's why I removed it from the build script and the before-commit hook (...not nice I know, but I wanted to see the build script completing).~
The typedoc script should work if [a PR](https://github.com/twilio/typedoc-twilio-theme/pull/3) on `twilio/typedoc-twilio-theme` will be accepted. 

Issue: https://github.com/twilio/twilio-client.js/issues/36#issuecomment-516071919

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [ ] Manually sanity tested running locally
* [ ] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

**Comment by giacomorebonato** [comment]
Please let me know what to do with this PR.
I am OK to improve it or close it, following your directions :)

**Comment by charliesantos** [comment]
Great addition. Thank you @giacomorebonato . We will review and let you know if additional action is required.

**Comment by vpodolyan** [comment]
We're using `twilio-client` in our TypesScript project, it'd be really great to make use of types :)

File: package.json [hunk 1]
```diff
@@ -107,7 +107,7 @@
     "typedoc-plugin-external-module-name": "2.1.0",
     "typedoc-plugin-internal-external": "2.0.2",
     "typedoc-twilio-theme": "1.0.0",
-    "typescript": "3.5.3",
+    "typescript": "^3.7.4",
```
Comment (ryan-rowland) [inline package.json]: Can you remove the caret (`^`) here and above to match the rest of the file?

File: package.json [hunk 2]
```diff
@@ -15,11 +16,11 @@
     "url": "git@github.com:twilio/twilio-client.js.git"
   },
   "scripts": {
-    "build": "npm-run-all clean build:version build:errors docs:ts build:es5 build:ts build:dist build:dist-min",
+    "build": "npm-run-all clean build:version build:errors build:es5 build:ts build:dist build:dist-min",
     "build:errors": "node ./scripts/errors.js",
     "build:es5": "rimraf ./es5 && babel lib -d es5",
     "build:dev": "ENV=dev npm run build",
-    "build:dist": "node ./scripts/build.js ./lib/browser.js ./LICENSE.md ./dist/twilio.js",
+    "build:dist": "node ./scripts/build.js ./es5/browser.js ./LICENSE.md ./dist/twilio.js",
```
Comment (ryan-rowland) [inline package.json]: Why is this change necessary? Was something breaking for you in browser?

Comment (giacomorebonato) [reply package.json]: I did this because I converted `browser.js` to TypeScript and then the output of the `node ./scripts/build.js` was containing TypeScript markup.
For the moment I put back `browser.js` as it is on master and revert this script as it was.

Having `browser.js` instead of `.ts` shouldn't impact on the production of typings.

**Comment by giacomorebonato** [comment]
> Looks good overall, left a couple comments and I'd rather not pull something that breaks our build process (docs necessary). If you're having trouble with some dependency issues with the typedoc generation, it's likely due to your twilio-node and typescript version updates. Maybe best to roll them back and we can handle that update separately.

If I roll back TypeScript version to **3.5.3** I fall into [this issue](https://github.com/angular/angular/issues/27921) where:
>Error: error TS5053: Option 'allowJs' cannot be specified with option 'declaration'.

So `allowJs` blocks me from outputting the type declaration files.
Do you have any suggestions about this?

**Comment by giacomorebonato** [comment]
And this is what happens with TypeDoc after updating TypeScript:

```bash
❯ npm run docs:ts

> twilio-client@1.9.8-dev docs:ts /Users/grebonato/Code/twilio-client.js
> typedoc --out docs --internal-aliases internal,publicapi --external-aliases external,internalapi --excludePrivate --excludeProtected --theme ./node_modules/typedoc-twilio-theme/bin/default

Loaded plugin /Users/grebonato/Code/twilio-client.js/node_modules/typedoc-plugin-external-module-name
Loaded plugin /Users/grebonato/Code/twilio-client.js/node_modules/typedoc-plugin-internal-external

Using TypeScript 3.5.3 from /Users/grebonato/Code/twilio-client.js/node_modules/typedoc/node_modules/typescript/lib
Rendering [==--------------------------------------] 3%/Users/grebonato/Code/twilio-client.js/node_modules/typedoc-twilio-theme/bin/default/helpers/index.js:3
    return context.filter(ctx =>
                   ^

TypeError: Cannot read property 'filter' of undefined
    at DeclarationReflection.eachWithPublic (/Users/grebonato/Code/twilio-client.js/node_modules/typedoc-twilio-theme/bin/default/helpers/index.js:3:20)
    at Object.eval [as main] (eval at createFunctionContext (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:266:23), <anonymous>:5:112)
    at main (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/runtime.js:176:32)
    at ret (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/runtime.js:179:12)
    at ret (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/compiler/compiler.js:515:21)
    at Object.invokePartial (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/runtime.js:295:12)
    at Object.invokePartialWrapper [as invokePartial] (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/runtime.js:78:39)
    at eval (eval at createFunctionContext (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:266:23), <anonymous>:8:28)
    at prog (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/runtime.js:229:12)
    at PageEvent.<anonymous> (/Users/grebonato/Code/twilio-client.js/node_modules/handlebars/dist/cjs/handlebars/helpers/with.js:32:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! twilio-client@1.9.8-dev docs:ts: `typedoc --out docs --internal-aliases internal,publicapi --external-aliases external,internalapi --excludePrivate --excludeProtected --theme ./node_modules/typedoc-twilio-theme/bin/default`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the twilio-client@1.9.8-dev docs:ts script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/grebonato/.npm/_logs/2020-01-09T22_41_18_650Z-debug.log
```

**Comment by giacomorebonato** [comment]
This change on typedoc-twilio-theme allows to complete the build: https://github.com/twilio/typedoc-twilio-theme/pull/3

**Comment by giacomorebonato** [comment]
Hi, since the PR on typedoc-twilio-theme has been merged, may I ask you to publish the updated version **1.0.1** to npm please? :)

**Comment by ryan-rowland** [comment]
Thanks again for your patience @giacomorebonato 

I've gone ahead and updated the npm to 1.0.1

**Comment by giacomorebonato** [comment]
No worries, there's no rush for this feature :)
I like to use this library every day for work and I am glad to be taken into consideration for contributing to it.

Now the build completes successfully, but on Travis there are failure related to:
```
Error: accountSid is required
```

I'll look into it in the next days.
Cheers.

**Comment by giacomorebonato** [comment]
I think that I need your support here @ryan-rowland 
I don't understand:
* how to execute `npm run test:integration` on my dev machine?
* why Travis is failing for missing env var?

**Comment by ryan-rowland** [comment]
Hmm, the accountSid issue is odd. The accountSid secret is stored in Travis so unless it's something that was introduced in the PR I'm not quite sure why that happened.

For the integration test you'll need to create a credentials file and set up an appropriate TwiML application. See the testing section of the readme: https://github.com/twilio/twilio-client.js#testing

**Comment by giacomorebonato** [comment]
I am able to run the tests locally now, but on Travis I can't tell why ENV vars are missing in the image:

Screenshot 2020-02-03 at 10 43 05

**Comment by giacomorebonato** [comment]
I am not too practical with Travis, but maybe this problem raises because the PR is coming from a fork: https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings

**Comment by giacomorebonato** [comment]
Any updates on this?

**Comment by giacomorebonato** [comment]
My current development is published on npm as `giacomo-twilio-client`.
It works when imported with the following syntax:

```ts
import * as twilio from 'giacomo-twilio-client'

twilio.Connection
twilio.Device
twilio.PStream
```

**Comment by charliesantos** [comment]
Thanks @giacomorebonato . We'll review and perform some tests on this in the next 1-2 weeks. We'll keep you updated.

**Comment by giacomorebonato** [comment]
Thanks! :)

**Comment by charliesantos** [comment]
Thank you @giacomorebonato for this PR. We are now tracking this here https://github.com/twilio/twilio-client.js/pull/109.
